documentation/Defining-Services.md

```diff
@@ -0,0 +1,171 @@
+# Defining Services
+
+Trapperkeeper provides two constructs for defining services: `defservice` and `service`.  As you might expect, `defservice` defines a service as a var in your namespace, and `service` allows you to create one inline and assign it to a variable in a let block or other location.  Here's how they work:
+
+## `defservice`
+
+`defservice` takes the following arguments:
+
+* a service name
+* an optional doc string
+* an optional service protocol; only required if your service exports functions that can be used by other services
+* a dependency list indicating other services/functions that this service requires
+* a series of function implementations.  This must include all of the functions in the protocol if one is specified, and may also optionally provide override implementations for the built-in service `Lifecycle` functions.
+
+### Service Lifecycle
+
+The service `Lifecycle` protocol looks like this:
+
+```clj
+(defprotocol Lifecycle
+  (init [this context])
+  (start [this context])
+  (stop [this context]))
+```
+
+(This may look familiar; we chose to use the same function names as some of the existing lifecycle protocols.  Ultimately we'd like to just use one of those protocols directly, but for now our needs are different enough to warrant avoiding the introduction of a dependency on an existing project.)
+
+All service lifecycle functions are passed a service `context` map, which may be used to store any service-specific state (e.g., a database connection pool or some other object that you need to reference in subsequent functions.)  Services may define these functions, `assoc` data into the map as needed, and then return the updated context map.  The updated context map will be maintained by the framework and passed to subsequent lifecycle functions for the service.
+
+The default implementation of the lifecycle functions is to simply return the service context map unmodified; if you don't need to implement a particular lifecycle function for your service, you can simply omit it and the default will be used.
+
+Trapperkeeper will call the lifecycle functions in order based on the dependency list of the services; in other words, if your service has a dependency on service `Foo`, you are guaranteed that `Foo`'s `init` function will be called prior to yours, and that your `stop` function will be called prior to `Foo`'s.
+
+### Example Service
+
+Let's look at a concrete example:
+
+```clj
+;; This is the list of functions that the `FooService` must implement, and which
+;; are available to other services who have a dependency on `FooService`.
+(defprotocol FooService
+  (foo1 [this x])
+  (foo2 [this])
+  (foo3 [this x]))
+
+(defservice foo-service
+   ;; docstring (optional.md)
```

Comment:
This might want to link to https://github.com/justinstoller/trapperkeeper/blob/task/TK-325/move-wiki-to-docs/documentation/Defining-Services.md#optional-services, or lose the .md.